### PR TITLE
Birmingham | ITP Jan-2026 | Arunkumar Akilan | Sprint 3 | Implement and Rewrite Tests

### PR DESCRIPTION
**Pull request by arunkumarakilan**
### Self checklist

* [x] I have titled my PR with Region | Cohort | FirstName LastName | Sprint | Assignment Title
* [x] My changes meet the requirements of the task
* [x] I have tested my changes
* [x] My changes follow the style guide

### Changelist

Implemented the required functions and updated the test files for the Sprint 3 **Implement and Rewrite Tests** exercises.

Completed the following:

* Implemented `getAngleType`
* Implemented `isProperFraction`
* Implemented `getCardValue`
* Rewrote the corresponding tests using **Jest**
* Ensured all tests pass successfully

### Questions

No questions at this time.

### CODE REVIEW

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js [hunk 1]
```diff
@@ -11,7 +11,18 @@
...
 function isProperFraction(numerator, denominator) {
-  // TODO: Implement this function
+  if (denominator === 0){
+    return false;
+  }
+  else if (numerator < denominator)
+  {
+    return true;
+
+  }
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js]: Can you lookup if -1/-2, 1/-2, -1/2 are considered proper fractions, and then update 
your implementation and tests accordingly?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 2]
```diff
@@ -9,9 +9,28 @@ const getAngleType = require("../implement/1-get-angle-type");
 test(`should return "Acute angle" when (0 < angle < 90)`, () => {
   // Test various acute angles, including boundary cases
   expect(getAngleType(1)).toEqual("Acute angle");
-  expect(getAngleType(45)).toEqual("Acute angle");
+  //expect(getAngleType(45)).toEqual("Acute angle");
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: Why commented out this test? If it is not needed, we should delete it to keep the code clean.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 3]
```diff
@@ -9,9 +9,28 @@ const getAngleType = require("../implement/1-get-angle-type");
...
   expect(getAngleType(89)).toEqual("Acute angle");
 });
+test(`should return "Right angle" when (angle = 90)`, () => {
+  // Test various acute angles, including boundary cases
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: This comment is not quite accurate.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js [hunk 4]
```diff
@@ -9,9 +9,28 @@ const getAngleType = require("../implement/1-get-angle-type");
...
+  expect(getAngleType(180)).toEqual("Straight angle");
+});
+test(`should return "Reflex angle" when (180 < angle <360)`, ()=>{
+  expect(getAngleType(181)).toEqual("Reflex angle");
+  expect(getAngleType(359)).toEqual("Reflex angle");
+
+});
+test(`should return "invalid angle when angles outside the valid range`, ()=>{
+  expect(getAngleType(0)).toEqual("Invalid angle")
+})
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: - Would be helpful to clearly indicate what constitutes the valid range in the description.

- Why not test both boundary cases and more samples in this category?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js [hunk 5]
```diff
@@ -8,7 +8,27 @@ const getCardValue = require("../implement/3-get-card-value");
...
 });
-
+test(`should return 9 when given a 9♠ card`,()=>{
+  expect(getCardValue("9♠")).toEqual(9);
+});
+test (`should return 10 when given a face card K♦`, ()=>{
+  expect(getCardValue("K♦")).toEqual(10);
+});
+test(`should return 10 when given a 10`, ()=>{
+  expect(getCardValue("10♥")).toEqual(10);
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js]: When preparing tests, we should ensure the tests cover all possible cases. If we specify a test for individual card, we will need about 53 tests to cover all possible cases. Instead, we could consider classifying all possible values into different categories, and then within each category we test some samples.

For example, one possible category for `getCardValue()` is, "should return the numeric value of number cards (2-10)", and we can prepare the test as 

```
test("should return the value of number cards (2-10)", () => {
    expect(getCardValue("2♣︎")).toEqual(2);
    expect(getCardValue("5♠")).toEqual(5);
    expect(getCardValue("10♥")).toEqual(10);

    // Note: We could also use a loop to check all values from 2 to 10.
});
```

**Can you practice preparing tests in this fashion?**

**Review by cjyuan — COMMENTED** [review]
Indentation is off and some of the code is not consistently formatted.

Have you installed the **prettier** VSCode extension and enabled "Format on save/paste" on VSCode,
as recommended in 
https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/blob/main/readme.md
?

---
If you have enabled "Format on save" but it is not working, it is likely that you haven't assign a formatter for JS file. This could happen if you have zero or multiple extensions that can format .js file.

If you have installed "Prettier" extension. To assign it as the formatter of JS code, you can try:
1. Use "Format document" to format the JS file. Sometimes, VSCode will ask you to choose a formatter, and you can manually select "Prettier".
2. Edit `settings.json` and set Prettier as the default formatter for JS.
See: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

**Comment by arunkumarakilan** [comment]
> Indentation is off and some of the code is not consistently formatted.
> 
> Have you installed the **prettier** VSCode extension and enabled "Format on save/paste" on VSCode, as recommended in https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/blob/main/readme.md ?
> 
> If you have enabled "Format on save" but it is not working, it is likely that you haven't assign a formatter for JS file. This could happen if you have zero or multiple extensions that can format .js file.
> 
> If you have installed "Prettier" extension. To assign it as the formatter of JS code, you can try:
> 
> 1. Use "Format document" to format the JS file. Sometimes, VSCode will ask you to choose a formatter, and you can manually select "Prettier".
> 2. Edit `settings.json` and set Prettier as the default formatter for JS.
>    See: https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode

Thanks for the feedback. I have now installed the Prettier extension in VS Code, enabled format on save and paste, and formatted the files. The indentation and formatting should now be consistent. Please let me know if anything else needs improvement.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js (hunk 2 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: Thanks for the suggestion. I have removed the commented test to keep the code clean.

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js (hunk 1 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js]: Thanks for the suggestion. I checked the definition of proper fractions and updated my implementation to compare absolute values. I also added test cases for -1/2, 1/-2 and -1/-2.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js (hunk 3 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: You're right — that comment was inaccurate for the right-angle test. I've removed it to keep the code clean. Thanks for pointing it out.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js (hunk 4 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/1-get-angle-type.test.js]: Thanks for the suggestion. I updated the test to clearly indicate the valid range and added more invalid cases such as negative angles and values above the valid range.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js (hunk 5 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/3-get-card-value.test.js]: Thanks for the suggestion. I updated the tests to group card values into categories (number cards 2–10, face cards, and Ace) and added tests for invalid inputs such as invalid ranks, suits, missing suits, and non-string inputs.

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js [hunk 6]
```diff
@@ -11,7 +11,11 @@
 // execute the code to ensure all tests pass.
 
 function isProperFraction(numerator, denominator) {
-  // TODO: Implement this function
+  if (denominator === 0) {
+    return false;
+  } {
+      return Math.abs(numerator) < Math.abs(denominator);
+  }
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js:16]: Code is still correct but looks a bit unusual. Is this intended?

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js [hunk 7]
```diff
@@ -8,3 +8,24 @@ const isProperFraction = require("../implement/2-is-proper-fraction");
...
+});
+test(`should return true when denominator is greater than numerator`, () => {
+  expect(isProperFraction(1, 2)).toEqual(true);
+});
+test(`should return false when numerator is greater than denominator`, () => {
+  expect(isProperFraction(5, 4)).toEqual(false);
+});
+test(`should return false when numerator is equals to denominator`, () => {
+  expect(isProperFraction(3, 3)).toEqual(false);
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js:22]: This test for boundary case is good.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js [hunk 8]
```diff
@@ -8,3 +8,24 @@ const isProperFraction = require("../implement/2-is-proper-fraction");
...
+test(`should return true when denominator is greater than negative numerator`, () => {
+  expect(isProperFraction(-1, 2)).toEqual(true);
+});
+test(`should return false when numerator is greater than denominator`, () => {
+  expect(isProperFraction(1, -2)).toEqual(true);
+});
+test(`should return false when negative numerator is greater than negative denominator`, () => {
+  expect(isProperFraction(-1, -2)).toEqual(true);
+});
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: These descriptions are not quite correct.

It is ok to use pseudo-code such as `abs(...)` and relational operators in the description if they can make the description more concise.

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js (hunk 6 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js:16]: Good catch. The extra block was unnecessary. I simplified the function to return the comparison directly after the denominator check.

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js (hunk 7 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js:22]: thank you

File: Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js (hunk 8 above)
Comment (arunkumarakilan) [reply Sprint-3/1-implement-and-rewrite-tests/rewrite-tests-with-jest/2-is-proper-fraction.test.js]: Good point. The descriptions didn’t clearly reflect that the implementation compares `abs(numerator)` and `abs(denominator)`. I’ve updated the test descriptions to describe the rule using `abs(numerator) < abs(denominator)`.

File: Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js [hunk 9]
```diff
@@ -11,7 +11,11 @@
 // execute the code to ensure all tests pass.
 
 function isProperFraction(numerator, denominator) {
-  // TODO: Implement this function
+  if (denominator === 0) {
+    return false;
+  } 
+      return Math.abs(numerator) < Math.abs(denominator);
```
Comment (cjyuan) [inline Sprint-3/1-implement-and-rewrite-tests/implement/2-is-proper-fraction.js:17]: The indentation is off. Looks like you may not have successfully assigned "prettier" as your default JS formatter.

**Review by cjyuan — COMMENTED** [review]
Changes look good.

**Comment by illicitonion** [comment]
Closing PR because the January ITP run has finished. Feel free to re-open if you're still working on it.
